File: chatosdk/src/main/java/com/gamatechno/chato/sdk/app/chatrooms/ChatRoomsFragment.java
package com.gamatechno.chato.sdk.app.chatrooms;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.gamatechno.chato.sdk.R;
import com.gamatechno.chato.sdk.app.chatroom.ChatRoomActivity;
import com.gamatechno.chato.sdk.app.chatrooms.adapter.RoomAdapter;
import com.gamatechno.chato.sdk.app.chatrooms.helper.ChatRoomsHelper;
import com.gamatechno.chato.sdk.app.chatrooms.uimodel.ChatRoomsUiModel;
import com.gamatechno.chato.sdk.app.chatrooms.viewmodel.ChatRoomsViewModel;
import com.gamatechno.chato.sdk.data.constant.StringConstant;
import com.gamatechno.chato.sdk.utils.animation.AnimationToggle;
import com.gamatechno.ggfw.utils.GGFWUtil;
import com.gamatechno.ggfw.utils.RecyclerScroll;
import com.google.firebase.iid.FirebaseInstanceId;

import java.util.ArrayList;
import java.util.List;

import static com.gamatechno.chato.sdk.module.request.GGFWRest.CODE_NETWORKERROR;
import static com.gamatechno.chato.sdk.module.request.GGFWRest.CODE_NOCONNECTIONERROR;
import static com.gamatechno.chato.sdk.module.request.GGFWRest.CODE_SERVERERROR;

public class ChatRoomsFragment extends Fragment implements ChatRoomsView.View {

    private static final String TAG = ChatRoomsFragment.class.getSimpleName();

    RecyclerView rv;

    RelativeLayout helper_noconversation;

    RelativeLayout helper_noconnection;

    RelativeLayout helper_servererror;

    Button btn_tryconnection;

    Button btn_tryservererror;

    AnimationToggle helper_loading_top;

    SwipeRefreshLayout swipe;

    TextView tv_startnewconversation;

    boolean isGroup = false;
    RoomAdapter roomAdapter;

    ChatRoomsPresenter obrolanPresenter;

    ChatRoomsViewModel viewModel;

    boolean isLoadMore = true;
    IntentFilter filter;

    String keyword = "";

    LinearLayoutManager linearLayoutManager;

    private final BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String action = intent.getAction();

            Log.d(TAG, "Receiving....");
            if(action.equals(StringConstant.broadcast_refresh_chat)){
                obrolanPresenter.requestObrolan(true, keyword);
            }
        }
    };

    private void registerReceiver(){
        filter = new IntentFilter();
        filter.addAction(StringConstant.broadcast_refresh_chat);
    }

    public static ChatRoomsFragment newInstance(boolean isGroup, ChatRoomsViewModel viewModel) {
        ChatRoomsFragment fragment = new ChatRoomsFragment();
        fragment.isGroup = isGroup;
        fragment.viewModel = viewModel;
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.activity_chat_rooms_fragment, container, false);
        initView(rootView);

        linearLayoutManager = new LinearLayoutManager(getContext());
        rv.setLayoutManager(linearLayoutManager);
        helper_loading_top.setInOutAnimation(R.anim.pull_in_bottom, R.anim.push_out_top);

        registerReceiver();
        initObrolan();

        rv.addOnScrollListener(new RecyclerScroll(linearLayoutManager){

            @Override
            public void show() {
//                Log.d("ChatRoomsFragment", "show: ");
                viewModel.updateScrollStatus(true);
            }

            @Override
            public void hide() {
                viewModel.updateScrollStatus(false);
            }

            @Override
            public void loadMore() {
                if(obrolanPresenter.isSuccess && !obrolanPresenter.isLoading){
                    obrolanPresenter.requestObrolan(false, "");
                }
            }
        });

        tv_startnewconversation.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                viewModel.updateStartChat(true);
            }
        });

        swipe.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                keyword = "";
                obrolanPresenter.requestObrolan(true, "");
            }
        });

        observeViewModel();

        btn_tryservererror.setOnClickListener(onClickListener());
        btn_tryconnection.setOnClickListener(onClickListener());
        return rootView;
    }

    private void observeViewModel(){
        viewModel.initRequestDelete().observe(getViewLifecycleOwner(), new Observer<Boolean>() {
            @Override
            public void onChanged(@Nullable Boolean aBoolean) {
                if(aBoolean){
                    if(roomAdapter.getData() != null){
                        ChatRoomsUiModel model = ChatRoomsHelper.getChatRoomClicked(roomAdapter.getData());
                        obrolanPresenter.deleteRoom(model);
                        uncheckTheChatRoom(false);
                    }
                }
            }
        });

        viewModel.initKeyword().observe(getViewLifecycleOwner(), new Observer<String>() {
            @Override
            public void onChanged(@Nullable String s) {
                if(s.equals("")){
                    keyword = s;
                    uncheckTheChatRoom(false);
                }
            }
        });

        viewModel.initRequestPin().observe(getViewLifecycleOwner(), new Observer<Boolean>() {
            @Override
            public void onChanged(@Nullable Boolean isRoom) {
                if(isRoom != null){
                    if(roomAdapter.getData() != null){
                        obrolanPresenter.pinChatRoom(ChatRoomsHelper.getChatRoomClicked(roomAdapter.getData()), ChatRoomsHelper.totalPinnedChatRoom(roomAdapter.getData()));
                        uncheckTheChatRoom(false);
                    }
                }
            }
        });


        viewModel.initChatRoomClickFromSearch().observe(getViewLifecycleOwner(), new Observer<ChatRoomsUiModel>() {
            @Override
            public void onChanged(@Nullable ChatRoomsUiModel model) {
                if(ChatRoomsHelper.getChatRoomClicked(roomAdapter.getData()) != null){
                    if(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), model) != -1){
                        if(ChatRoomsHelper.getChatRoomClicked(roomAdapter.getData()).getRoomChat().getRoom_id() == model.getRoomChat().getRoom_id()){
                            roomAdapter.getData().get(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), model)).setIs_checked(false);
                        } else {
                            checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), model), true);
                        }
                        viewModel.updateChatRoomsLongPress(model);
                    }
                } else {
                    Intent intent = new Intent(getContext(), ChatRoomActivity.class);
                    intent.putExtra("chatroom", model);
                    startActivity(intent);
                }
                roomAdapter.notifyDataSetChanged();
            }
        });

        viewModel.initChatRoomLongPressFromSearch().observe(getViewLifecycleOwner(), new Observer<ChatRoomsUiModel>() {
            @Override
            public void onChanged(@Nullable ChatRoomsUiModel chatRoomUiModel) {
                if(ChatRoomsHelper.getChatRoomClicked(roomAdapter.getData()) != null){
                    if(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), chatRoomUiModel) != -1){
                        if(ChatRoomsHelper.getChatRoomClicked(roomAdapter.getData()).getRoomChat().getRoom_id() == chatRoomUiModel.getRoomChat().getRoom_id()){
                            roomAdapter.getData().get(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), chatRoomUiModel)).setIs_checked(false);
                        } else {
                            checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), chatRoomUiModel), true);
                        }
                        viewModel.updateChatRoomsLongPress(chatRoomUiModel);
                    }
                } else {
                    if(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), chatRoomUiModel) != -1){
                        checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), chatRoomUiModel), true);
                        viewModel.updateChatRoomsLongPress(chatRoomUiModel);
                    }
                }
                roomAdapter.notifyDataSetChanged();
            }
        });
        viewModel.initRefreshRoom().observe(getViewLifecycleOwner(), new Observer<Boolean>() {
            @Override
            public void onChanged(@Nullable Boolean aBoolean) {
                obrolanPresenter.requestObrolan(true, "");
            }
        });
    }

    public View.OnClickListener onClickListener() {
        return new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(view.getId() == R.id.btn_tryservererror || view.getId() == R.id.btn_tryconnection){
                    swipe.setRefreshing(false);
                    keyword = "";
                    obrolanPresenter.requestObrolan(true, "");
                }
            }
        };
    }

    private void initObrolan(){
        obrolanPresenter = new ChatRoomsPresenter(getContext(), this);
        roomAdapter = new RoomAdapter(getContext(), new RoomAdapter.OnObrolanAdapter() {
            @Override
            public void onClickObrolan(ChatRoomsUiModel model) {
                if(ChatRoomsHelper.getChatRoomClicked(roomAdapter.getData()) != null){
                    if(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), model) != -1){
                        if(ChatRoomsHelper.getChatRoomClicked(roomAdapter.getData()).getRoomChat().getRoom_id() == model.getRoomChat().getRoom_id()){
                            roomAdapter.getData().get(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), model)).setIs_checked(false);
                        } else {
                            checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), model), true);
                        }
                        viewModel.updateChatRoomsLongPress(model);
                    }
                } else {
                    Intent intent = new Intent(getContext(), ChatRoomActivity.class);
                    intent.putExtra("chatroom", model);
                    startActivity(intent);
                }
                roomAdapter.notifyDataSetChanged();
            }

            @Override
            public void onLongClick(ChatRoomsUiModel chatRoomUiModel) {
                if(ChatRoomsHelper.getChatRoomClicked(roomAdapter.getData()) != null){
                    if(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), chatRoomUiModel) != -1){
                        if(ChatRoomsHelper.getChatRoomClicked(roomAdapter.getData()).getRoomChat().getRoom_id() == chatRoomUiModel.getRoomChat().getRoom_id()){
                            roomAdapter.getData().get(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), chatRoomUiModel)).setIs_checked(false);
                        } else {
                            checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), chatRoomUiModel), true);
                        }
                        viewModel.updateChatRoomsLongPress(chatRoomUiModel);
                    }
                } else {
                    if(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), chatRoomUiModel) != -1){
                        checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter.getData(), chatRoomUiModel), true);
                        viewModel.updateChatRoomsLongPress(chatRoomUiModel);
                    }
                }
                roomAdapter.notifyDataSetChanged();
            }
        });
        rv.setAdapter(roomAdapter);
    }

    private void checkTheChatRoom(int i, boolean isOK){
        for (int j = 0; j < roomAdapter.getData().size(); j++) {
            if(j == i){
                roomAdapter.getData().get(j).setIs_checked(isOK);
            } else {
                roomAdapter.getData().get(j).setIs_checked(!isOK);
            }
        }
    }

    private void uncheckTheChatRoom(boolean isOk){
        for (int j = 0; j < roomAdapter.getData().size(); j++) {
            roomAdapter.getData().get(j).setIs_checked(isOk);
        }
        roomAdapter.notifyDataSetChanged();
    }

    @Override
    public void onLoading() {
        helper_noconversation.setVisibility(View.GONE);
        helper_noconnection.setVisibility(View.GONE);
        helper_servererror.setVisibility(View.GONE);
        if(roomAdapter.getData().size() == 0){
            roomAdapter.initLoading(true);
        }
    }

    @Override
    public void onHideLoading() {
        helper_loading_top.hide();
        if(roomAdapter.getData().size() == 0){
            roomAdapter.initLoading(false);
        } else {
            if(swipe.isRefreshing()) swipe.setRefreshing(false);
        }
    }

    @Override
    public void onErrorConnection(String message) {
        if(roomAdapter.getData().size()==0){
            switch (message){
                case CODE_SERVERERROR:
                    helper_servererror.setVisibility(View.VISIBLE);
                    break;
                case CODE_NETWORKERROR:
                    helper_noconnection.setVisibility(View.VISIBLE);
                    break;
                case CODE_NOCONNECTIONERROR:
                    helper_noconnection.setVisibility(View.VISIBLE);
                    break;
                default:
                    GGFWUtil.ToastShort(getContext(), message);
                    break;
            }
        } else {
            GGFWUtil.ToastShort(getContext(), message);
        }
    }

    @Override
    public void onAuthFailed(String error) {
        GGFWUtil.ToastShort(getContext(), error);
    }

    @Override
    public void onRequestObrolan(List<ChatRoomsUiModel> list, boolean isRefresh) {
        isLoadMore = true;
        helper_noconversation.setVisibility(View.GONE);
        roomAdapter.addData(isRefresh, list);
        if(roomAdapter.getData().size() == 0)
            helper_noconversation.setVisibility(View.VISIBLE);
        else
            helper_noconversation.setVisibility(View.GONE);
    }

    @Override
    public void onFailedRequestObrolan() {
        if(roomAdapter.getData().size() == 0){
            helper_noconversation.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void successPinnedChatRoom(String message) {
        GGFWUtil.ToastShort(getContext(), message);
        obrolanPresenter.requestObrolan(true, keyword);
    }

    @Override
    public void onDeleteRoom(boolean isSuccess, String message) {
        GGFWUtil.ToastShort(getContext(), message);
        if(isSuccess){
            obrolanPresenter.requestObrolan(true, keyword);
        }
    }

    @Override
    public void onStop() {
        try {
            if(getContext()!=null)
                getContext().unregisterReceiver(receiver);
        } catch (Exception e){
            e.getMessage();
        }
        super.onStop();
    }

    @Override
    public void onResume() {
        super.onResume();
        Log.d(TAG, "onResume: "+FirebaseInstanceId.getInstance().getToken());
        getContext().registerReceiver(receiver, filter);
        obrolanPresenter.requestObrolan(true, keyword);
    }

    private void initView(View view){
        rv = view.findViewById(R.id.rv);
        helper_noconversation = view.findViewById(R.id.helper_noconversation);
        helper_noconnection = view.findViewById(R.id.helper_noconnection);
        helper_servererror = view.findViewById(R.id.helper_servererror);
        btn_tryconnection = view.findViewById(R.id.btn_tryconnection);
        btn_tryservererror = view.findViewById(R.id.btn_tryservererror);
        helper_loading_top = view.findViewById(R.id.helper_loading_top);
        swipe = view.findViewById(R.id.swipe);
        tv_startnewconversation = view.findViewById(R.id.tv_startnewconversation);
    }
}

File: chatosdk/src/main/java/com/gamatechno/chato/sdk/app/chatrooms/ChatRoomsFragment.kt
package com.gamatechno.chato.sdk.app.chatrooms

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.gamatechno.chato.sdk.R
import com.gamatechno.chato.sdk.app.chatroom.ChatRoomActivity
import com.gamatechno.chato.sdk.app.chatrooms.adapter.RoomAdapter
import com.gamatechno.chato.sdk.app.chatrooms.adapter.RoomAdapter.OnObrolanAdapter
import com.gamatechno.chato.sdk.app.chatrooms.adapter.RoomFilterAdapter
import com.gamatechno.chato.sdk.app.chatrooms.helper.ChatRoomsHelper
import com.gamatechno.chato.sdk.app.chatrooms.uimodel.ChatRoomsUiModel
import com.gamatechno.chato.sdk.app.chatrooms.viewmodel.ChatRoomsViewModel
import com.gamatechno.chato.sdk.data.DAO.Label.LabelModel
import com.gamatechno.chato.sdk.data.constant.StringConstant
import com.gamatechno.chato.sdk.module.request.GGFWRest
import com.gamatechno.ggfw.utils.GGFWUtil
import com.gamatechno.ggfw.utils.RecyclerScroll
import com.google.firebase.iid.FirebaseInstanceId
import kotlinx.android.synthetic.main.activity_chat_rooms_fragment.*
import kotlinx.android.synthetic.main.layout_helper.*

class ChatRoomsFragment : Fragment(), ChatRoomsView.View {
    var isGroup = false
    var roomAdapter: RoomAdapter? = null
    var obrolanPresenter: ChatRoomsPresenter? = null
    var viewModel: ChatRoomsViewModel? = null
    var isLoadMore = true
    var filter: IntentFilter? = null
    var keyword = ""
    var needShowLoadingView = false
    lateinit var filter_adapter : RoomFilterAdapter
    var linearLayoutManager: LinearLayoutManager? = null
    private val receiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action
            Log.d(TAG, "Receiving....")
            if (action == StringConstant.broadcast_refresh_chat) {
                obrolanPresenter!!.requestObrolan(true, keyword, filter_adapter.getSortBy())
            }
        }
    }

    private fun registerReceiver() {
        filter = IntentFilter()
        filter!!.addAction(StringConstant.broadcast_refresh_chat)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        linearLayoutManager = LinearLayoutManager(context)
        rv.layoutManager = linearLayoutManager

        filter_adapter = RoomFilterAdapter(requireContext(), object : RoomFilterAdapter.OnRoomFilter{
            override fun onSetRoomFilter(list: MutableList<LabelModel>) {
                needShowLoadingView = true
                obrolanPresenter!!.requestObrolan(true, keyword, filter_adapter.getSortBy())
            }
        })
        filter_adapter.setDatas(ChatRoomsHelper.filtered_labels())
        rv_filter.adapter = filter_adapter

        helper_loading_top!!.setInOutAnimation(R.anim.pull_in_bottom, R.anim.push_out_top)
        registerReceiver()
        initObrolan()
        rv!!.addOnScrollListener(object : RecyclerScroll(linearLayoutManager) {
            override fun show() {
//                Log.d("ChatRoomsFragment", "show: ");
                viewModel!!.updateScrollStatus(true)
            }

            override fun hide() {
                viewModel!!.updateScrollStatus(false)
            }

            override fun loadMore() {
                if (obrolanPresenter!!.isSuccess && !obrolanPresenter!!.isLoading) {
                    obrolanPresenter!!.requestObrolan(false, "", filter_adapter.getSortBy())
                }
            }
        })
        tv_startnewconversation!!.setOnClickListener { viewModel!!.updateStartChat(true) }
        swipe!!.setOnRefreshListener {
            keyword = ""
            obrolanPresenter!!.requestObrolan(true, "", filter_adapter.getSortBy())
        }
        observeViewModel()
        btn_tryservererror!!.setOnClickListener(onClickListener())
        btn_tryconnection!!.setOnClickListener(onClickListener())
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.activity_chat_rooms_fragment, container, false)
        return rootView
    }

    private fun observeViewModel() {
        viewModel!!.initRequestDelete().observe(viewLifecycleOwner, Observer { aBoolean ->
            if (aBoolean!!) {
                if (roomAdapter!!.data != null) {
                    val model = ChatRoomsHelper.getChatRoomClicked(roomAdapter!!.data)
                    obrolanPresenter!!.deleteRoom(model)
                    uncheckTheChatRoom(false)
                }
            }
        })
        viewModel!!.initKeyword().observe(viewLifecycleOwner, Observer { s ->
            if (s == "") {
                keyword = s
                uncheckTheChatRoom(false)
            }
        })
        viewModel!!.initRequestPin().observe(viewLifecycleOwner, Observer { isRoom ->
            if (isRoom != null) {
                if (roomAdapter!!.data != null) {
                    obrolanPresenter!!.pinChatRoom(ChatRoomsHelper.getChatRoomClicked(roomAdapter!!.data), ChatRoomsHelper.totalPinnedChatRoom(roomAdapter!!.data))
                    uncheckTheChatRoom(false)
                }
            }
        })
        viewModel!!.initChatRoomClickFromSearch().observe(viewLifecycleOwner, Observer { model ->
            if (ChatRoomsHelper.getChatRoomClicked(roomAdapter!!.data) != null) {
                if (ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, model) != -1) {
                    if (ChatRoomsHelper.getChatRoomClicked(roomAdapter!!.data)?.roomChat?.room_id == model!!.roomChat.room_id) {
                        roomAdapter!!.data[ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, model)].is_checked = false
                    } else {
                        checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, model), true)
                    }
                    viewModel!!.updateChatRoomsLongPress(model)
                }
            } else {
                val intent = Intent(context, ChatRoomActivity::class.java)
                intent.putExtra("chatroom", model)
                startActivity(intent)
            }
            roomAdapter!!.notifyDataSetChanged()
        })
        viewModel!!.initChatRoomLongPressFromSearch().observe(viewLifecycleOwner, Observer { chatRoomUiModel ->
            if (ChatRoomsHelper.getChatRoomClicked(roomAdapter!!.data) != null) {
                if (ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, chatRoomUiModel) != -1) {
                    if (ChatRoomsHelper.getChatRoomClicked(roomAdapter!!.data)?.roomChat?.room_id == chatRoomUiModel!!.roomChat.room_id) {
                        roomAdapter!!.data[ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, chatRoomUiModel)].is_checked = false
                    } else {
                        checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, chatRoomUiModel), true)
                    }
                    viewModel!!.updateChatRoomsLongPress(chatRoomUiModel)
                }
            } else {
                if (ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, chatRoomUiModel) != -1) {
                    checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, chatRoomUiModel), true)
                    viewModel!!.updateChatRoomsLongPress(chatRoomUiModel)
                }
            }
            roomAdapter!!.notifyDataSetChanged()
        })
        viewModel!!.initRefreshRoom().observe(viewLifecycleOwner, Observer { obrolanPresenter!!.requestObrolan(true, "", filter_adapter.getSortBy()) })
    }

    fun onClickListener(): View.OnClickListener {
        return View.OnClickListener { view ->
            if (view.id == R.id.btn_tryservererror || view.id == R.id.btn_tryconnection) {
                swipe!!.isRefreshing = false
                keyword = ""
                obrolanPresenter!!.requestObrolan(true, "", filter_adapter.getSortBy())
            }
        }
    }

    private fun initObrolan() {
        obrolanPresenter = ChatRoomsPresenter(context, this)
        roomAdapter = RoomAdapter(context, object : OnObrolanAdapter {
            override fun onClickObrolan(model: ChatRoomsUiModel) {
                if (ChatRoomsHelper.getChatRoomClicked(roomAdapter!!.data) != null) {
                    if (ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, model) != -1) {
                        if (ChatRoomsHelper.getChatRoomClicked(roomAdapter!!.data)?.roomChat?.room_id == model.roomChat.room_id) {
                            roomAdapter!!.data[ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, model)].is_checked = false
                        } else {
                            checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, model), true)
                        }
                        viewModel!!.updateChatRoomsLongPress(model)
                    }
                } else {
                    val intent = Intent(context, ChatRoomActivity::class.java)
                    intent.putExtra("chatroom", model)
                    startActivity(intent)
                }
                roomAdapter!!.notifyDataSetChanged()
            }

            override fun onLongClick(chatRoomUiModel: ChatRoomsUiModel) {
                if (ChatRoomsHelper.getChatRoomClicked(roomAdapter!!.data) != null) {
                    if (ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, chatRoomUiModel) != -1) {
                        if (ChatRoomsHelper.getChatRoomClicked(roomAdapter!!.data)?.roomChat?.room_id == chatRoomUiModel.roomChat.room_id) {
                            roomAdapter!!.data[ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, chatRoomUiModel)].is_checked = false
                        } else {
                            checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, chatRoomUiModel), true)
                        }
                        viewModel!!.updateChatRoomsLongPress(chatRoomUiModel)
                    }
                } else {
                    if (ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, chatRoomUiModel) != -1) {
                        checkTheChatRoom(ChatRoomsHelper.getIndexChatRoom(roomAdapter!!.data, chatRoomUiModel), true)
                        viewModel!!.updateChatRoomsLongPress(chatRoomUiModel)
                    }
                }
                roomAdapter!!.notifyDataSetChanged()
            }
        })
        rv!!.adapter = roomAdapter
    }

    private fun checkTheChatRoom(i: Int, isOK: Boolean) {
        for (j in roomAdapter!!.data.indices) {
            if (j == i) {
                roomAdapter!!.data[j].is_checked = isOK
            } else {
                roomAdapter!!.data[j].is_checked = !isOK
            }
        }
    }

    private fun uncheckTheChatRoom(isOk: Boolean) {
        for (j in roomAdapter!!.data.indices) {
            roomAdapter!!.data[j].is_checked = isOk
        }
        roomAdapter!!.notifyDataSetChanged()
    }

    override fun onLoading() {
        helper_noconversation!!.visibility = View.GONE
        helper_noconnection!!.visibility = View.GONE
        helper_servererror!!.visibility = View.GONE
        if (roomAdapter!!.data.size == 0) {
            roomAdapter!!.initLoading(true)
        } else {
            if(needShowLoadingView){
                swipe!!.isRefreshing = false
                roomAdapter!!.initLoading(true)
            }
        }
    }

    override fun onHideLoading() {
        helper_loading_top!!.hide()
        if (roomAdapter!!.data.size == 0 || needShowLoadingView) {
            needShowLoadingView = false
            roomAdapter!!.initLoading(false)
        } else {
            if (swipe!!.isRefreshing) swipe!!.isRefreshing = false
        }
    }

    override fun onErrorConnection(message: String) {
        if (roomAdapter!!.data.size == 0) {
            when (message) {
                GGFWRest.CODE_SERVERERROR -> helper_servererror!!.visibility = View.VISIBLE
                GGFWRest.CODE_NETWORKERROR -> helper_noconnection!!.visibility = View.VISIBLE
                GGFWRest.CODE_NOCONNECTIONERROR -> helper_noconnection!!.visibility = View.VISIBLE
                else -> GGFWUtil.ToastShort(context, message)
            }
        } else {
            GGFWUtil.ToastShort(context, message)
        }
    }

    override fun onAuthFailed(error: String) {
        GGFWUtil.ToastShort(context, error)
    }


    override fun onRequestObrolan(list: List<ChatRoomsUiModel?>?, isRefresh: Boolean) {
        isLoadMore = true
        helper_noconversation!!.visibility = View.GONE
        roomAdapter!!.addData(isRefresh, list)
        if (roomAdapter!!.data.size == 0) helper_noconversation!!.visibility = View.VISIBLE else helper_noconversation!!.visibility = View.GONE
    }

    override fun successPinnedChatRoom(message: String?) {
        GGFWUtil.ToastShort(context, message)
        obrolanPresenter!!.requestObrolan(true, keyword, filter_adapter.getSortBy())
    }

    override fun onDeleteRoom(isSuccess: Boolean, message: String?) {
        GGFWUtil.ToastShort(context, message)
        if (isSuccess) {
            obrolanPresenter!!.requestObrolan(true, keyword, filter_adapter.getSortBy())
        }
    }

    override fun onFailedRequestObrolan() {
        if (roomAdapter!!.data.size == 0 || needShowLoadingView) {
            helper_noconversation!!.visibility = View.VISIBLE
            needShowLoadingView = false
        }
    }


    override fun onStop() {
        try {
            if (context != null) context!!.unregisterReceiver(receiver)
        } catch (e: Exception) {
            e.message
        }
        super.onStop()
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "onResume: " + FirebaseInstanceId.getInstance().token)
        context!!.registerReceiver(receiver, filter)
        obrolanPresenter!!.requestObrolan(true, keyword, filter_adapter.getSortBy())
    }

    companion object {
        private val TAG = ChatRoomsFragment::class.java.simpleName
        @JvmStatic
        fun newInstance(isGroup: Boolean, viewModel: ChatRoomsViewModel): ChatRoomsFragment {
            val fragment = ChatRoomsFragment()
            fragment.isGroup = isGroup
            fragment.viewModel = viewModel
            return fragment
        }
    }
}